presto-main/src/main/java/com/facebook/presto/resourcemanager/LeaderResourceManagerAddressSelector.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.resourcemanager;
+
+import com.facebook.drift.client.address.AddressSelector;
+import com.facebook.drift.client.address.SimpleAddressSelector.SimpleAddress;
+import com.facebook.presto.metadata.InternalNode;
+import com.facebook.presto.metadata.InternalNodeManager;
+import com.facebook.presto.spi.HostAddress;
+import com.google.common.net.HostAndPort;
+
+import javax.inject.Inject;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class LeaderResourceManagerAddressSelector
+        implements AddressSelector<SimpleAddress>
+{
+    private final InternalNodeManager internalNodeManager;
+    private final RatisClient client;
+
+    @Inject
+    public LeaderResourceManagerAddressSelector(InternalNodeManager internalNodeManager, RatisClient ratisClient)
+    {
+        requireNonNull(internalNodeManager, "internalNodeManager is null");
+        this.internalNodeManager = internalNodeManager;
+        requireNonNull(ratisClient, "ratisClient is null");
+        this.client = ratisClient;
+    }
+
+    @Override
+    public Optional<SimpleAddress> selectAddress(Optional<String> addressSelectionContext)
+    {
+        if (addressSelectionContext.isPresent()) {
+            return addressSelectionContext
+                    .map(HostAndPort::fromString)
+                    .map(SimpleAddress::new);
+        }
+
+        for (InternalNode resourceManager : internalNodeManager.getResourceManagers()) {
+            if (resourceManager.getNodeIdentifier().equals(client.getLeader())) {
```

Comment:
It does reach out to the network every time but the only reason I didn't capture it was because the leader could change due to a leader election so the string wouldn't be updated if I do capture it.